### DISCUSSION TITLE
docs: Opt out of "Unit of Work caching"

### DISCUSSION THREAD

**Discussion by zetti-caletti** [mikro-orm/mikro-orm#4135] [General]
As @TypeOrm is sadly a dead project now, I was looking for other ORMs in the nodejs world and found Mikroorm.
In the docs I saw, that Mikroorm tries to return the entity from the cache (on second run) when queried by their primary key.
As nodejs are, compared to for example php scripts, long running processes, how does this behave if I have multiple nodejs processes/containers/servers running? If one process updates an entity and an other process/server queries for that entity and has it in his cache, then I am getting an outdated result. In the docs it says that you can specify your custom ResultCacheAdapter. I havent found a build in for Redis/Memcached, etc. Do I have to implement that by myself? Or is it best to completely disable the result cache (is possible?).
I found that MikroOrm has a RedisCacheAdapter, but it is listed under Metadata Cache, so it will not be used for Result Cache?
Thanks for answer and maybe the docs can be improved here a little bit.  :)

**Comment by B4nan** [comment]
> As nodejs are, compared to for example php scripts, long running processes, how does this behave if I have multiple nodejs processes/containers/servers running?

You are asking for more docs and it feels like you haven't read the current ones :]

https://mikro-orm.io/docs/installation#request-context
https://mikro-orm.io/docs/entity-manager#disabling-identity-map-and-change-set-tracking
https://mikro-orm.io/docs/identity-map#how-does-requestcontext-helper-work

> In the docs it says that you can specify your custom ResultCacheAdapter

Feels like you are mixing identity map, result cache, and metadata cache, those are 3 different things. And yes, you need to bring your own cache adapters, MikroORM provides only one for memory (used for result cache) and one for JSON files (used for metadata cache). You can't configure the "identity map cache", that always lives in the memory, as that is tied to the EntityManager instance, you can't share it, it would break the change tracking. Note that the result cache is opt-in, you need to enable that on the call site, it is not enabled by default.

> I found that MikroOrm has a RedisCacheAdapter, but it is listed under Metadata Cache, so it will not be used for Result Cache?

It does not have a `RedisCacheAdapter`, that is only an example of how you can create your own adapter. That class does not exist anywhere outside the docs.

**Reply by zetti-caletti** [reply]
I still didn't got it. According to this:

> MikroORM uses the Identity Map pattern to track objects. Whenever you fetch an object from the database, MikroORM will keep a reference to this object inside its UnitOfWork.
> This allows MikroORM room for optimizations. If you call the EntityManager and ask for an entity with a specific ID twice, it will return the same instance:
> 
> const authorRepository = orm.em.getRepository(Author);
> const jon1 = await authorRepository.findOne(1);
> const jon2 = await authorRepository.findOne(1);
> 
> // identity map in action
> console.log(jon1 === jon2); // true
> 
> Only one SELECT query will be fired against the database here. In the second findOne() call MikroORM will check the identity map first and will skip the database round trip as it will find the entity already loaded.

Here I haven't opt-in to caching.
So as far as I understand, if I do this:
```
const jon1 = await authorRepository.findOne(1);

// after above call, other process deletes author with id 1

setTimeout(() => {
    const jon2 = await authorRepository.findOne(1);
}, 500);
```

Then the call in the setTimeout returns an entity, although it doesn't exist in the database anymore?

And according to the docs:

> MikroORM has simple result caching mechanism. It works with those methods of EntityManager: find(), findOne(), findAndCount(), findOneOrFail(), count(), as well as with QueryBuilder result methods (including execute).

ResultCache only caches results from EntityManager and QueryBuilder calls, but not from Repository calls? Does this makes any sense?

And no, I read the docs, but don't get the logic yet^^

**Reply by zetti-caletti** [reply]
> Entity Repositories are thin layers on top of EntityManager

So does ResultCache also covers calls to EntityRepository? If yes, then this should be included in the docs of ResultCache

**Reply by zetti-caletti** [reply]
Regarding RequestContext:
So for every HTTP-Request, my backend receives, I should create a new RequestContext, and every RequestContext has its own cache?
Why do I then have a cache at all? Why should I call "const jon = await authorRepository.findOne(1);" twice in the same request?

**Reply by zetti-caletti** [reply]
[https://medium.com/dailyjs/mikroorm-4-1-lets-talk-about-performance-b184d612f1d5](https://medium.com/dailyjs/mikroorm-4-1-lets-talk-about-performance-b184d612f1d5)

> Iyibo Eki
> Thanks for working on this.
My concern with Mikro-ORM and the Unit of work approach are these:
1- What is the the impact to memory, given the added unit-of-work layer?
2- What is the performance across multiple orm forks? as opposed to just performance within a single fork as this test focuses on?
Remember, a realistic use case of this ORM is that the unit of Work context is forked upon every single request made to a server. Can we see more information around these common use cases?
Thanks for the great work!

**Reply by zetti-caletti** [reply]
I mean, whats the purpose of caching in Unit of Work if I should create a new cache (RequestContext) for every HTTP-Request?
Would be much better to have a shared/distributed cache like Redis (with user limited max amount of memory) and every process can use this cache to reduce database calls. And then remove the caching in Unit of Work completely.

**Reply by zetti-caletti** [reply]
Imagine we have two servers and a shared network cache:

Server 1 becomes a request:
/api/posts/1

Mikroorm always looks first in shared network cache (for example redis). There is not entry for Post with id 1. So it makes a database call and after that writes the result to the shared network cache.

Server 2 becomes a request:
/api/posts/1
Mikroorm always looks first in shared network cache and sees that there is an entity and returns that entity from the cache. 

If an entity is deleted or updated, the entity in the shared network cache has to be updated as well.

**Reply by zetti-caletti** [reply]
The cache could look like this:

key: {tableName}_{primaryKey}
value: serialized entity
```
key: 'post_1'
value: {
    id: 1,
    name: 'Post One'
}

key: 'book_ea463a60-0f98-4464-97f9-af398ffc6fd8'
value: {
    id: 'ea463a60-0f98-4464-97f9-af398ffc6fd8',
    name: 'Book One'
}
```

**Reply by zetti-caletti** [reply]
I mean the idea of caching the things is very nice. But as far as I understand the whole cachiing mechanism only works if I have only one server/process running. If more, then things get broken, because the cache is local in memory and not shared.

**Reply by B4nan** [reply]
Please read those pages, again, more carefully, you still confuse identity map with result cache, those are two different things, with a different purpose. An identity map is supposed to be never shared, that is the whole point of it, as it guarantees only a single instance of a given entity exists in the context (usually a single request).

> ResultCache only caches results from EntityManager and QueryBuilder calls, but not from Repository calls? Does this makes any sense?

Don't read too much between the lines. If you would open the repository docs, or even better, the code, you would see what a repository does - it only wraps EM. It works exactly the same, as it uses the EM to do everything.

> But as far as I understand the whole cachiing mechanism only works if I have only one server/process running. If more, then things get broken, because the cache is local in memory and not shared.

And that is where you should put your redis cache adapter. For result cache it is indeed good idea. 

> I mean, whats the purpose of caching in Unit of Work if I should create a new cache (RequestContext) for every HTTP-Request?

To have single instance of a given entity in one context - to protect you from doing updates to a single entity (by its PK) multiple times, overriding each other. Don't think about it as a cache, it is not a cache per-se, it deduplicates entity instances to protect you from mistakes. It is rather a cache for entity instances, not a cache for querying database (that is exactly result cache).
